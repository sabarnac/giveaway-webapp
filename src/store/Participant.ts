import { observable, computed } from "mobx";

/**
 * Class containing the details of a participants' avatar.
 */
export class Avatar {
  @observable private _url: string;
  @observable private _altText: string;

  public constructor(url: string, altText: string) {
    this._url = url;
    this._altText = altText;
  }

  /**
   * Gets the avatar URL.
   * @return {string} The avatar URL.
   */
  public get url(): string {
    return this._url;
  }

  /**
   * Gets the avatar image alternate text.
   * @return {string} The avatar image alternate text.
   */
  public get altText(): string {
    return this._altText;
  }

  private isSameUrl = (url1: string, url2: string): boolean => url1 === url2;

  private isSameAltText = (altText1: string, altText2: string): boolean =>
    altText1 === altText2;

  /**
   * Determines whether another avatar is identical to the current one.
   * @param  {Avatar} otherAvatar The other avatar to compare against.
   * @returns {boolean} Whether the other avatar is equal to the current one.
   */
  public equals = (otherAvatar: Avatar): boolean =>
    this.isSameUrl(this._url, otherAvatar._url) &&
    this.isSameAltText(this._altText, otherAvatar._altText);
}

/**
 * Class containing the details of a participant.
 */
export default class Participant {
  @observable private _name: string;
  @observable private _avatar?: Avatar;

  public constructor(name: string, avatar?: Avatar) {
    this._name = name;
    this._avatar = avatar;
  }

  /**
   * Gets the participant name.
   * @return {string} The participant name.
   */
  public get name(): string {
    return this._name;
  }

  /**
   * Gets the participant avatar, if it exists.
   * @return {string | null} The participant avatar, or null if there isn't one.
   */
  public get avatar(): Avatar | null {
    return this._avatar || null;
  }

  /**
   * Gets whether the participant has an avatar or not.
   * @returns {boolean} Whether the participant has an avatar or not.
   */
  @computed public get hasAvatar(): boolean {
    return !!this._avatar;
  }

  /**
   * Gets the capitalized name of the participant.
   * @returns {string} The capitalized name.
   */
  @computed public get properName(): string {
    return this._name.replace(/\w\S*/g, (subText: string) => {
      return subText.charAt(0).toUpperCase() + subText.substr(1).toLowerCase();
    });
  }

  private isSameName = (name1: string, name2: string): boolean =>
    name1 === name2;

  private isSameAvatar = (avatar1?: Avatar, avatar2?: Avatar): boolean =>
    avatar1 && avatar2
      ? avatar1.equals(avatar2)
      : !avatar1 && !avatar2
      ? true
      : false;

  /**
   * Determines whether another participant is identical to the current one.
   * @param  {Participant} otherParticipant The other participant to compare against.
   * @returns {boolean} Whether the other participant is equal to the current one.
   */
  public equals = (otherParticipant: Participant): boolean =>
    this.isSameName(this._name, otherParticipant._name) &&
    this.isSameAvatar(this._avatar, otherParticipant._avatar);
}
