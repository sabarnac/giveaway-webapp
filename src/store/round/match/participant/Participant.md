Usage:

```tsx
import Avatar from "./store/round/match/participant/Avatar";

const participant: Participant = new Participant(
  "FOO BAR" /*, {url:"url",altText:"alt-text"} */
);
const participantName: string = participant.name; // "FOO BAR"
const participantProperName: string = participant.properName; // "Foo Bar"
const participantAvatar: Avatar = participant.avatar; // Avatar("url","alt-text")
```
