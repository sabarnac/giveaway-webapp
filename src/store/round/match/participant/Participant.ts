import { observable, computed } from "mobx";
import { createAvatarImage } from "../../../../util";
import Avatar from "./Avatar";
const inflect = require("i")();

interface AvatarJson {
  url: string;
  altText: string;
}

/**
 * Class containing the details of a participant.
 */
export default class Participant {
  /** The ID of the participant. */
  @observable private _id: string;
  /** The name of the participant. */
  @observable private _name: string;
  /** The avatar of the participant. */
  @observable private _avatar: Avatar;
  /** The weight of the participant. */
  @observable private _weight: number;

  /** A counter for generating a unique ID for the match. */
  private static counter: number = 1;

  /**
   * Gets the avatar of the participant if present, or creates a random one,
   * @param  {string} name The name of the participant.
   * @param  {AvatarJson} avatar The avatar details of the participant, if available.
   * @return {Avatar} The generated avatar of the participant.
   */
  private _getOrCreateAvatar = (name: string, avatar?: AvatarJson): Avatar => {
    return avatar
      ? new Avatar(avatar.url, avatar.altText)
      : new Avatar(createAvatarImage(name), name);
  };

  public constructor(name: string, avatar?: AvatarJson, weight: number = 1) {
    this._id = `${Participant.counter++}`;
    this._name = name;
    this._avatar = this._getOrCreateAvatar(name, avatar);
    this._weight = weight;
  }

  /**
   * Get the participant ID.
   * @return {string} The unique ID of the participant.
   */
  @computed public get id(): string {
    return this._id;
  }

  /**
   * Gets the participant name.
   * @return {string} The participant name.
   */
  @computed public get name(): string {
    return this._name;
  }

  /**
   * Gets the capitalized name of the participant.
   * @return {string} The capitalized name.
   */
  @computed public get properName(): string {
    return inflect.titleize(this._name);
  }

  /**
   * Gets the participant avatar, if it exists.
   * @return {string} The participant avatar.
   */
  @computed public get avatar(): Avatar {
    return this._avatar;
  }

  /**
   * Gets the participant weight.
   * @return {number} The participant weight.
   */
  @computed public get weight(): number {
    return this._weight;
  }

  /**
   * Returns whether the two names are equal.
   * @param {string} name1 The first name to compare.
   * @param {string} name2 The second name to compare.
   * @return {boolean} Whether the two names are equal.
   */
  private _isSameName = (name1: string, name2: string): boolean =>
    name1 === name2;

  /**
   * Returns whether the two avatars are equal.
   * @param {Avatar} avatar1 The first avatar to compare.
   * @param {Avatar} avatar2 The second avatar to compare.
   * @return {boolean} Whether the two avatars are equal.
   */
  private _isSameAvatar = (avatar1: Avatar, avatar2: Avatar): boolean =>
    avatar1.equals(avatar2);

  /**
   * Determines whether another participant is identical to the current one.
   * @param  {Participant} otherParticipant The other participant to compare against.
   * @return {boolean} Whether the other participant is equal to the current one.
   */
  public equals = (otherParticipant: Participant): boolean =>
    this._isSameName(this._name, otherParticipant._name) &&
    this._isSameAvatar(this._avatar, otherParticipant._avatar);
}
