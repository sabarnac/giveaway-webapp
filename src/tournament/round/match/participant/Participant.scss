.participant {
    background: #ffffff;
    &--invert {
        background: #000000;
        color: #ffffff;
    }
}