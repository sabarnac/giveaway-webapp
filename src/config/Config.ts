import Participant, { Avatar } from "../store/Participant";
import ConfigJson from "./config.json";
import Random, { Engine } from "random-js";

/**
 * Interface for JSON object representing avatar details.
 */
interface AvatarJson {
  url: string;
  altText: string;
}

/**
 * Interface for JSON object representing participant details.
 */
interface ParticipantJson {
  name: string;
  avatar?: AvatarJson;
}

/**
 * Enumeration of possible animation speeds.
 */
export const enum AnimationSpeed {
  HALF = 0.5,
  ONE = 1,
  ONE_POINT_FIVE = 1.5,
  TWO = 2,
  FIVE = 5,
  TEN = 10,
  TWENTY = 20,
}

/**
 * Class representing the basic configuration of the application.
 */
class Config {
  private _randomGenerator: Random;
  private _messages: string[];
  private _allParticipants: Participant[];
  private _speed: AnimationSpeed;
  private _participantsPerMatch: AnimationSpeed;

  private getRandomGenerator = (): Random => {
    const randomEngine: Engine = Random.engines.mt19937();
    return new Random(randomEngine);
  };

  private getMessage = (): string[] => ConfigJson.messages;

  private createParticipant = (user: string | ParticipantJson): Participant =>
    typeof user === "string"
      ? new Participant(user)
      : new Participant(
          (<ParticipantJson>user).name,
          this.createAvatar((<ParticipantJson>user).avatar),
        );

  private createAvatar = (avatar?: AvatarJson): Avatar | undefined =>
    avatar ? new Avatar(avatar.url, avatar.altText) : undefined;

  private getParticipants = (): Participant[] =>
    ConfigJson.users.map(this.createParticipant);

  private getSpeed = (): AnimationSpeed => AnimationSpeed.ONE;

  private getParticipantsPerMatch = (): number =>
    ConfigJson.participantsPerMatch;

  constructor() {
    this._randomGenerator = this.getRandomGenerator();
    this._messages = this.getMessage();
    this._allParticipants = this.getParticipants();
    this._speed = this.getSpeed();
    this._participantsPerMatch = this.getParticipantsPerMatch();
  }

  /**
   * Get the random generator.
   * @return {Random}
   */
  public get randomGenerator(): Random {
    return this._randomGenerator;
  }

  /**
   * Get the list of winner/loser messages.
   * @return {string[]}
   */
  public get messages(): string[] {
    return this._messages;
  }

  /**
   * Get the list of all participants in the giveaway.
   * @return {Participant[]}
   */
  public get allParticipants(): Participant[] {
    return this._allParticipants;
  }

  /**
   * Get the speed multiplier for animations.
   * @return {AnimationSpeed}
   */
  public get speed(): AnimationSpeed {
    return this._speed;
  }

  /**
   * Set the speed multiplier for animations.
   * @param {AnimationSpeed} value The new speed multiplier value.
   */
  public set speed(value: AnimationSpeed) {
    this._speed = value;
  }

  /**
   * Get the number of participants per match.
   * @return {AnimationSpeed}
   */
  public get participantsPerMatch(): AnimationSpeed {
    return this._participantsPerMatch;
  }
}

export default new Config();
