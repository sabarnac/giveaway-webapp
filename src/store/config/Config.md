Assume the configuration in `src/config/config.json` is set as:

```json
{
  "name": "Foobar Giveaway",
  "seed": 1234567890,
  "participantsPerMatch": 2,
  "messages": [
    "#winner #loser 1",
    "#winner #loser 2",
    "#winner #loser 3",
    "#winner #loser 4"
  ],
  "users": ["FOO", "BAR", "FAZ"]
}
```

Usage:

```tsx
import AnimationSpeed from "./store/config/AnimationSpeed";
import Participant from "./store/round/match/participant/Participant";

const name: string = Config.name; // "Foobar Giveaway
const messages: string[] = Config.messages; // ["#winner #loser 1", "#winner #loser 2", "#winner #loser 3", "#winner #loser 4"]
const message: string = Config.getRandomMessage("foobar", ["barfoo"]); // "foobar barfoo 3"
const allParticipants: Participant[] = Config.allParticipants; // [Participant("FOO"), Participant("BAR"), Participant("FAZ")]
const participantsPerMatch: number = Config.participantsPerMatch; // 2
const speed: number = Config.speed; // 1
Config.setSpeed(AnimationSpeed.TWO);
```
