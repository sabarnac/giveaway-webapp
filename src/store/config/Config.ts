import Participant from "../round/match/participant/Participant";
import ConfigJson from "./config.json";
import { observable, computed, action } from "mobx";
import RandomGenerator from "./RandomGenerator";
import AnimationSpeed from "./AnimationSpeed";

/**
 * Interface for JSON object representing avatar details.
 */
interface AvatarJson {
  /** The URL of the avatar. */
  url: string;
  /** The alternate text of the avatar image. */
  altText: string;
}

/**
 * Interface for JSON object representing participant details.
 */
interface ParticipantJson {
  /** The name of the participant. */
  name: string;
  /** The avatar of the participant (if present). */
  avatar?: AvatarJson;
  /** The weight of the participant (if present). */
  weight?: number;
}

/**
 * Class representing the basic configuration of the application.
 */
export default class Config {
  /** The name of the tournament. */
  private _name: string;
  /** A list of all messages. */
  private _messages: string[];
  /** The list of participants in the tournament. */
  private _allParticipants: Participant[];
  /** The number of participants per match. */
  private _participantsPerMatch: number;
  /** The animation speed multiplier. */
  @observable private _speed: number;
  /** The list of unused messages. */
  private _unusedMessages: string[];

  /** The singleton instance of the class, or null if not yet created. */
  private static _instance: Config | null = null;

  /**
   * Returns the name of the tournament.
   * @return {string} The tournament name.
   */
  private _getName = (): string => ConfigJson.name;

  /**
   * Returns the match conclusion messages that can be used for the tournament.
   * @return {string[]} The tournament match conclusion messages.
   */
  private _getMessages = (): string[] => ConfigJson.messages;

  /**
   * Returns a created participant for the given JSON details.
   * @param {string | ParticipantJson} user The participants details.
   * @return {Participant} The created participant.
   */
  private _createParticipant = (user: string | ParticipantJson): Participant =>
    typeof user === "string"
      ? new Participant(user)
      : new Participant(
          (<ParticipantJson>user).name,
          (<ParticipantJson>user).avatar,
          (<ParticipantJson>user).weight,
        );

  /**
   * Returns a created list of participants using the given participant details.
   * @return {Participant[]} The created participants list.
   */
  private _getParticipants = (): Participant[] =>
    ConfigJson.users.map(this._createParticipant);

  /**
   * Returns the max. number of participants that are to be present per match in the tournament.
   * @return {number} The max. number of participants per match.
   */
  private _getParticipantsPerMatch = (): number =>
    ConfigJson.participantsPerMatch;

  /**
   * Returns the animation speed multiplier of the application.
   * @return {number} The animation speed multiplier.
   */
  private _getSpeed = (): number => AnimationSpeed.get("ONE") as number;

  /**
   * Returns whether a string should be prepended to a participants' name.
   * @param {number} index The index of the participant.
   * @param {number} participantSize The total number of participants.
   * @return {boolean} Whether the string should be prepended.
   */
  private _shouldPrepend = (index: number, participantSize: number): boolean =>
    index === participantSize - 1 && participantSize > 1;

  /**
   * Returns a function that can returns the participant name prepended with a string if it's the last participant, or just the name.
   * @param {number} participantSize The total number of participants.
   * @param {string} stringToPrepend The string to prepend.
   * @return {(name: string, index: number) => string} The function that will return the participants' name
   */
  private _prependToLastParticipant = (
    participantSize: number,
    stringToPrepend: string,
  ): ((name: string, index: number) => string) => (
    name: string,
    index: number,
  ): string =>
    this._shouldPrepend(index, participantSize)
      ? `${stringToPrepend} ${name}`
      : name;

  /**
   * Returns a formatted losers list.
   * @param {string[]} names The list of loser names.
   * @return {string} The formatted string.
   */
  private _formatLosers = (names: string[]): string =>
    names.map(this._prependToLastParticipant(names.length, "and")).join(", ");

  private constructor() {
    this._name = this._getName();
    this._messages = this._getMessages();
    this._unusedMessages = RandomGenerator.shuffle([...this._messages]);
    this._allParticipants = this._getParticipants();
    this._participantsPerMatch = this._getParticipantsPerMatch();
    this._speed = this._getSpeed();
  }

  /**
   * Get an instance of the config, which is a singleton.
   * @return {Config} The instance of a config.
   */
  public static getInstance(): Config {
    return Config._instance
      ? Config._instance
      : (Config._instance = new Config());
  }

  /**
   * Get the name of the tournament.
   * @return {string} The tournament name.
   */
  public get name(): string {
    return this._name;
  }

  /**
   * Get the list of winner/loser messages.
   * @return {string[]} The list of mesasges.
   */
  public get messages(): string[] {
    return [...this._messages];
  }

  /**
   * Returns the formatted version of the given message.
   * @param {string} message The message to format.
   * @param {string} winnerName The name of the winner.
   * @param {string[]} loserNames The names of the losers.
   * @return {string} The formatted message.
   */
  public getFormattedMessage = (
    message: string,
    winnerName: string,
    loserNames: string[],
  ): string =>
    message
      .replace("#winner", winnerName)
      .replace("#loser", this._formatLosers(loserNames));

  /**
   * Gets a random message with the winner and loser.
   * @return {string} The formatted message.
   */
  public getRandomMessage(): string {
    if (this._unusedMessages.length === 0) {
      this._unusedMessages = RandomGenerator.shuffle([...this._messages]);
    }
    const message = RandomGenerator.pick(this._unusedMessages);
    this._unusedMessages.splice(this._unusedMessages.indexOf(message), 1);
    return message;
  }

  /**
   * Gets the index of the given message.
   * @param {string} message The message.
   * @return {number} The index of the message.
   */
  public getMessageIndex = (message: string): number =>
    this._messages.indexOf(message);

  /**
   * Get the list of all participants in the tournament.
   * @return {Participant[]} The list of all participants.
   */
  public get allParticipants(): Participant[] {
    return this._allParticipants;
  }

  /**
   * Get the number of participants per match.
   * @return {number} The number of participants per match.
   */
  public get participantsPerMatch(): number {
    return this._participantsPerMatch;
  }

  /**
   * Get the speed multiplier for animations.
   * @return {number} The speed multiplier value.
   */
  @computed public get speed(): number {
    return this._speed;
  }

  /**
   * Set the speed multiplier for animations.
   * @param {number} value The new speed multiplier value.
   */
  @action public setSpeed(value: number): void {
    if (AnimationSpeed.hasValue(value)) {
      this._speed = value;
    } else {
      console.error(`Illegal value '${value}' for speed`);
    }
  }
}
