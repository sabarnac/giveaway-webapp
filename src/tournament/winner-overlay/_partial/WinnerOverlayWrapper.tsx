import React from "react";
import { observer, inject } from "mobx-react";
import "./WinnerOverlayWrapper.scss";
import classNames from "classnames";
import Participant from "../../../store/round/match/participant/Participant";
import Config from "../../../store/config/Config";
import WinnerOverlayView from "./WinnerOverlayView";

/**
 * Properties of the winner overlay wrapper React component.
 */
interface WinnerOverlayWrapperProps {
  /** @ignore The application config. */
  config?: Config;
  /** The tournament winner. */
  winner: Participant;
}

/**
 * React component for the winner overlay wrapper.
 */
export default inject("config")(
  observer(
    (props: WinnerOverlayWrapperProps): JSX.Element => (
      <div
        className={classNames("winner-overlay-wrapper")}
        style={{
          transition: `opacity ${500 / props.config!.speed}ms ease-in-out`,
        }}
      >
        <WinnerOverlayView winner={props.winner} />
      </div>
    ),
  ),
);
