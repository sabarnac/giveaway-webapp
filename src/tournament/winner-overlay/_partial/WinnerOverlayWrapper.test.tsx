import React from "react";
import WinnerOverlayWrapper from "./WinnerOverlayWrapper";
import { create, ReactTestRenderer } from "react-test-renderer";
import Participant from "../../../store/round/match/participant/Participant";
import { createDummyParticipant, createDummyConfig } from "../../../util/test";
import Config from "../../../store/config/Config";

describe("WinnerOverlayWrapper Component.", () => {
  it("Matches snapshot.", () => {
    const config: Config = createDummyConfig();
    const participant: Participant = createDummyParticipant(0);
    const component: ReactTestRenderer = create(
      <WinnerOverlayWrapper config={config} winner={participant} />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
