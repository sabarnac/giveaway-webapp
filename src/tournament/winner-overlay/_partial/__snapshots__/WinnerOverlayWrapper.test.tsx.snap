// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WinnerOverlayWrapper Component. Matches snapshot. 1`] = `
<div
  className="winner-overlay-wrapper"
  style={
    Object {
      "transition": "opacity 500ms ease-in-out",
    }
  }
>
  <div
    className="winner-overlay"
  >
    <div
      className="winner-overlay__winner"
    >
      <div
        className="participant participant-card"
      >
        <div
          className="participant-card__avatar"
        >
          <img
            alt="barfoo-0"
            src="foobar-0"
          />
        </div>
        <div
          className="participant-card__name"
        >
          <strong>
            Fozbaz 0
          </strong>
        </div>
      </div>
      <h3>
        Won The Giveaway!
      </h3>
    </div>
  </div>
</div>
`;
